docs/architecture/adr-046-module-params.md

```diff
@@ -0,0 +1,165 @@
+to state. In addition, modules must implement `sdk.Msg` message(s) and their
+respective handler(s) such that when a governance proposal passes, via the work
+done in [#9810](https://github.com/cosmos/cosmos-sdk/pull/9810), the `x/gov`
+module can execute those messages which would then change the respective
+parameters and any other state mutations necessary.
+
+Note, it is up to developers to decide how to structure their parameters and
+the respective `sdk.Msg` messages. Consider the parameters currently defined in
+`x/auth` using the `x/params` module for parameter management:
+
+```protobuf
+message Params {
+  uint64 max_memo_characters       = 1;
+  uint64 tx_sig_limit              = 2;
+  uint64 tx_size_cost_per_byte     = 3;
+  uint64 sig_verify_cost_ed25519   = 4;
+  uint64 sig_verify_cost_secp256k1 = 5;
+}
+```
+
+Developers can choose to either create a unique data structure for every field in
+`Params` or they can create a single `Params` structure as outlined above in the
+case of `x/auth`.
+
+In the former approach, a `sdk.Msg` would need to be created for every single
+field along with a handler. This can become burdensome if there are a lot of
+parameter fields. In the latter case, there is only a single data structure and
+thus only a single message handler, however, the message handler might have to be
+more sophisticated in that it might need to understand what parameters are being
+changed vs what parameters are untouched.
+
+Once the `sdk.Msg` types and corresponding handlers are defined, proposals can
+be made using the `x/gov` module that are authorized by the root `x/gov`
+module's account.
+
+Continuing to use `x/auth`, we demonstrate a more complete example:
+
+```go
+type Params struct {
+	MaxMemoCharacters      uint64
+	TxSigLimit             uint64
+	TxSizeCostPerByte      uint64
+	SigVerifyCostED25519   uint64
+	SigVerifyCostSecp256k1 uint64
+}
+
+type MsgUpdateParams struct {
+	MaxMemoCharacters      uint64
+	TxSigLimit             uint64
+	TxSizeCostPerByte      uint64
+	SigVerifyCostED25519   uint64
+	SigVerifyCostSecp256k1 uint64
+}
+
+type MsgUpdateParamsResponse struct {}
+
+func (ms msgServer) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
+  ctx := sdk.UnwrapSDKContext(goCtx)
+
+  // verification logic...
+
+  // persist params
+  params := ParamsFromMsg(msg)
+  ms.SaveParams(ctx, params)
+
+  return &types.MsgUpdateParamsResponse{}, nil
+}
+
+func ParamsFromMsg(msg *types.MsgUpdateParams) Params {
+  // ...
+}
+```
+
+## Consequences
+
+As a result of implementing the module parameter methodology, we gain the ability
+for module parameter changes to be stateful and extensible to fit nearly every
+application's use case. In addition, there will be significant gains in performance
+when it comes to reading and writing parameters from and to state, especially if
+a specific set of parameters are read on a consistent basis.
```

Comment:
```suggestion
application's use case. We will be able to emit events (and trigger hooks registered 
to that events using the work proposed in [even hooks](https://github.com/cosmos/cosmos-sdk/discussions/9656))
as well as call other Msg service methods or perform additional migration (as long 
as it will be possible to execute in one transaction).
In addition, there will be significant gains in performance when it comes to reading 
and writing parameters from and to state, especially if a specific set of parameters 
are read on a consistent basis.
```